src/lib/replication/repl_dev/raft_repl_dev.cpp

```diff
@@ -107,13 +107,79 @@ bool RaftReplDev::join_group() {
         m_msg_mgr.join_group(m_group_id, "homestore_replication",
                              std::dynamic_pointer_cast< nuraft_mesg::mesg_state_mgr >(shared_from_this()));
     if (!raft_result) {
-        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", boost::uuids::to_string(m_group_id),
-                      raft_result.error());
+        HS_DBG_ASSERT(false, "Unable to join the group_id={} with error={}", group_id_str(), raft_result.error());
         return false;
     }
     return true;
 }
 
+AsyncReplResult<> RaftReplDev::replace_member(replica_id_t member_out_uuid, replica_id_t member_in_uuid) {
+    LOGINFO("Replace member group_id={} member_out={} member_in={}", group_id_str(),
+            boost::uuids::to_string(member_out_uuid), boost::uuids::to_string(member_in_uuid));
+
+    // Step 1: Check if leader itself is requested to move out.
+    if (m_my_repl_id == member_out_uuid && m_my_repl_id == get_leader_id()) {
+        // If leader is the member requested to move out, then give up leadership and return error.
+        // Client will retry replace_member request to the new leader.
+        raft_server()->yield_leadership(true /* immediate */, -1 /* successor */);
+        LOGINFO("Replace member leader is the member_out so yield leadership");
+        return make_async_error<>(ReplServiceError::NOT_LEADER);
+    }
+
+    // Step 2. Add the new member.
+    return m_msg_mgr.add_member(m_group_id, member_in_uuid)
+        .via(&folly::InlineExecutor::instance())
+        .thenValue([this, member_in_uuid, member_out_uuid](auto&& e) -> AsyncReplResult<> {
+            // TODO Currently we ignore the cancelled, fix nuraft_mesg to not timeout
+            // when adding member. Member is added to cluster config until member syncs fully
```

Comment:
This is a nuraft behaviour. I had same doubt. This what he shared.
"""
* there are three servers: s1, s2, s3. So quorum size is 2. 
* let's say s3 is lagging behind, but still s1 can make a commit, by forming quorum {s1, s2}.
* now new member s4 joins the cluster. 
* if s4 joins the cluster immediately without catch-up, it is also lagging behind.
* since it joins the cluster immediately, now quorum size becomes 3. 
* now s1 cannot make a commit, no matter which member is in the quorum: {s1, s2, s3}, {s1, s2, s4}.
"""

Reply:
this scenario makes sense.
another question , do we have a plan to have a separate add_member and rem_member?

for example, if we have a very heavy read workload for one pg,  we want to add some new members to take over some read workload. so we only need add_memeber.

Reply:
@JacksonYao287 we dont want that as of now. We dont have the flexibility to control replication size of each PG in control plane.  Lets revisit if we have solid use case.

Reply:
@sanebay that makes sense from the raft point of view, but make it very tricky for cm->sm protocal.  Once leader SM reboot, what would be the state of the in-flight add_srv? will it continue till success or being discard?  who should take the responsibility to respond to CM? 

Is there a possibility that we can add new_member as learner and promoted to participant once it catches up?

Reply:
also for a failed/forgot add_srv, who should take the responsibility to reclaim space?   If CM retries with same PG to same SM I believe baseline resync will erase and re-do.  If CM attempt with other PG to this SM I am not sure how we handle this --- linking to the fixed pg size design @JacksonYao287

Reply:
> @sanebay that makes sense from the raft point of view, but make it very tricky for cm->sm protocal. Once leader SM reboot, what would be the state of the in-flight add_srv? will it continue till success or being discard? who should take the responsibility to respond to CM?
> 
> Is there a possibility that we can add new_member as learner and promoted to participant once it catches up?

add_srv internally does cluster config change with log entry. So it needs commit . Its treated as same append entry log. If committed and crashed, CM can send again and we return error that server already exists.  If it didnt commit, it will rollback once another leader selects. I havent explored the learner option.


> also for a failed/forgot add_srv, who should take the responsibility to reclaim space? 
No as mentioned, add_srv should fail with error SERVER_ALREADY_EXISTS . As per design, if new member failed after sometime, manual intervnetion is needed and a new PG move is needed(https://github.com/eBay/NuRaft/blob/master/src/handle_join_leave.cxx#L68)

Reply:
>also for a failed/forgot add_srv, who should take the responsibility to reclaim space? If CM retries with same PG to same SM I believe baseline resync will erase and re-do. If CM attempt with other PG to this SM I am not sure how we handle this

yes, a failed add_srv will leave some stale data in the disk of member_in , it is caused by the snapshot before the config change is committed.

I have gone through the raft code. I will add the solution in the doc